plugin/evm/validators_transitioner_test.go

```diff
@@ -0,0 +1,163 @@
+
+		got, err := validatorsTransitioner.ByEpoch(testEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+		assert.Equal(t, 10, calls)
+
+		// calls is expected to increase by 1 only because the cache would be used
+		got, err = validatorsTransitioner.ByEpoch(testEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+		assert.Equal(t, 11, calls)
+
+		mock.ByEpochFunc = func(uint64) (map[ids.ShortID]uint64, error) {
+			calls++
+			return providers, nil
+		}
+
+		got, err = validatorsTransitioner.ByEpoch(testEpoch + 1)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+		assert.Equal(t, 13, calls)
+	})
+
+	t.Run("nominal case 2: non-nil default validators to reach the end of the function", func(t *testing.T) {
+		t.Parallel()
+
+		testEpoch := uint64(10)
+		wantWeight := uint64(45)
+		providers := fakeProviders(testEpoch)
+		// here we give non-nil validators so that the later part of the ByEpoch() function can be reached by control and therefore tested
+		validators := fakeValidators(testEpoch)
+
+		mock := &testValidatorRetriever{
+			ByEpochFunc: func(uint64) (map[ids.ShortID]uint64, error) {
+				return providers, nil
+			},
+		}
+
+		validatorsTransitioner := NewValidatorsTransitioner(validators, mock)
+
+		got, err := validatorsTransitioner.ByEpoch(testEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+
+		var totalWeight uint64
+		for _, u := range got {
+			totalWeight += u
+		}
+		assert.Equal(t, wantWeight, totalWeight)
+	})
+
+	t.Run("epoch less than 1", func(t *testing.T) {
+		t.Parallel()
+
+		testEpoch := uint64(0)
+		validators := fakeProviders(testEpoch)
+
+		mock := &testValidatorRetriever{
+			ByEpochFunc: func(epoch uint64) (map[ids.ShortID]uint64, error) {
+				assert.Equal(t, testEpoch, epoch)
+				return nil, nil
+			},
+		}
+
+		validatorsTransitioner := NewValidatorsTransitioner(validators, mock)
+
+		got, err := validatorsTransitioner.ByEpoch(testEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, validators, got)
+	})
+
+	t.Run("case for non nil error", func(t *testing.T) {
+		t.Parallel()
+
+		mock := &testValidatorRetriever{
+			ByEpochFunc: func(uint64) (map[ids.ShortID]uint64, error) {
+				return nil, fmt.Errorf("dummy error")
+			},
+		}
+
+		validatorsTransitioner := NewValidatorsTransitioner(nil, mock)
+
+		_, err := validatorsTransitioner.ByEpoch(1)
+		require.Error(t, err)
+	})
+
+	t.Run("case for nil default validators", func(t *testing.T) {
+		t.Parallel()
+
+		epoch := uint64(10)
+		providers := fakeProviders(epoch)
+
+		mock := &testValidatorRetriever{
+			ByEpochFunc: func(epoch uint64) (map[ids.ShortID]uint64, error) {
+				return providers, nil
+			},
+		}
+
+		validatorsTransitioner := NewValidatorsTransitioner(nil, mock)
+
+		got, err := validatorsTransitioner.ByEpoch(epoch)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+	})
+}
+
+func fakeProviders(epoch uint64) map[ids.ShortID]uint64 {
```

Comment:
Inside tests/mocks/validators.go, there is a specific function `func RetrieverMock(numValidators, epochs int) ValidatorsTestData` that creates random validators set with epochs.
You can simply use it by giving `1` as epoch number and then use the 0th position of the map that is received, 
as data material used across these tests.
Or, use the hardcoded validators values inside `ValidatorsData`.
The functionality for adding an additional byte, in order to be 100% sure the data is not the same, can be run on any of the above results, and IMO it should also be inside tests/mocks/validators as a method implemented by `ValidatorsTestData`, as it already implements `Duplicate`.

Reply:
This is true on your branch, but it has not been merged yet. We can do a round of cleanup when it gets merged, to uniformize the tests and their use of mocked values 👍

Reply:
> This is true on your branch, but it has not been merged yet. We can do a round of cleanup when it gets merged, to uniformize the tests and their use of mocked values +1
Yes